tools/dependency/validate.py

```diff
@@ -122,28 +107,51 @@ def query_external_deps(self, *targets):
         Returns:
           A set of dependency identifiers that are reachable from targets.
         """
-        deps_query = 'deps(set({}))'.format(' '.join(targets))
+        # Get the smallest sets possible to fulfill the query!
+        deps_query = self._filtered_deps_query(targets)
+        exclude_deps = set()
+        if exclude:
+            exclude_query = self._filtered_deps_query(exclude)
+            deps_query = f'{deps_query} - {exclude_query}'
         try:
-            deps = subprocess.check_output(['bazel', 'query', deps_query],
-                                           stderr=subprocess.PIPE).decode().splitlines()
-        except subprocess.CalledProcessError as exc:
-            print(
-                f'Bazel query failed with error code {exc.returncode} and std error: {exc.stderr.decode()}'
-            )
-            raise exc
+            deps = self._deps_query(deps_query)
+            if deps and exclude:
+                # although the deps set is pre-filtered to exclude
+                # the excluded deps, we still need to fetch the exclude set
+                # again and remove any further matches, due to rev dep mangling.
+                # The exclude set could be pre-filtered further (ie only members
+                # of the revmap.values) at the cost of some additional complexity.
+                exclude_deps = self._deps_query(exclude_query)
+        except BazelQueryError as e:
+            print(f'Bazel query failed with error {e}')
+            raise e
+        return deps - exclude_deps
+
+    @lru_cache(maxsize=5)
+    def _deps_query(self, query_string):
+        return self._mangle_deps_set(query(query_string))
+
+    def _filtered_deps_query(self, targets):
+        return f'filter("^@.*//", deps(set({" ".join(targets)})))'
+
+    def _mangle_deps_set(self, deps):
         ext_deps = set()
         implied_untracked_deps = set()
         for d in deps:
-            match = BAZEL_QUERY_EXTERNAL_DEP_RE.match(d)
-            if match:
-                ext_dep = match.group(1)
-                if ext_dep in self._ignore_deps:
-                    continue
-                # If the dependency is untracked, add the source dependency that loaded
-                # it transitively.
-                if ext_dep in self._implied_untracked_deps_revmap:
-                    ext_dep = self._implied_untracked_deps_revmap[ext_dep]
-                ext_deps.add(ext_dep)
+            if not d:
+                continue
+            try:
+                ext_dep = BAZEL_QUERY_EXTERNAL_DEP_RE.match(d).group(1)
+            except:
+                print(f"BAD MATCH: {d}")
```

Comment:
just continue on no match

Reply:
Seems a pretty broad exception catch, can this be made more specific?

Reply:
checking the original code it was "look before you leap", so ill just get it to continue if `match` is `None`

Reply:
Is this just an assert condition?

Reply:
ive made this just do it what it did before